crates/qpy/src/py_methods.rs

```diff
@@ -0,0 +1,666 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2025
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+// Methods for QPY serialization working directly with Python-based data
+use binrw::Endian;
+use numpy::Complex64;
+use pyo3::IntoPyObjectExt;
+use pyo3::exceptions::{PyIOError, PyTypeError, PyValueError};
+use pyo3::intern;
+use pyo3::prelude::*;
+use pyo3::types::{
+    PyAny, PyComplex, PyDict, PyFloat, PyInt, PyIterator, PyList, PyString, PyTuple,
+};
+use qiskit_circuit::classical::expr::Expr;
+use std::io::Cursor;
+
+use qiskit_circuit::Clbit;
+use qiskit_circuit::bit::{PyClassicalRegister, PyClbit, ShareableClbit};
+use qiskit_circuit::circuit_data::CircuitData;
+use qiskit_circuit::classical;
+use qiskit_circuit::imports;
+use qiskit_circuit::operations::{Operation, OperationRef, StandardInstruction};
+use qiskit_circuit::packed_instruction::{PackedInstruction, PackedOperation};
+use qiskit_circuit::parameter::parameter_expression::{
+    PyParameter, PyParameterExpression, PyParameterVectorElement,
+};
+
+use uuid::Uuid;
+
+use crate::bytes::Bytes;
+use crate::formats;
+use crate::value::{
+    GenericValue, QPYWriteData, deserialize, modifier_types, pack_generic_value,
+    serialize_generic_value, tags, type_name,
+};
+use binrw::BinWrite;
+
+const UNITARY_GATE_CLASS_NAME: &str = "UnitaryGate";
+
+fn is_python_gate(py: Python, op: &PackedOperation, python_gate: &Bound<PyAny>) -> PyResult<bool> {
+    match op.view() {
+        OperationRef::Gate(pygate) => {
+            if pygate.gate.bind(py).is_instance(python_gate)? {
+                Ok(true)
+            } else {
+                Ok(false)
+            }
+        }
+        _ => Ok(false),
+    }
+}
+
+/// custom gates have unique UUID attached to their name
+/// this method recognizes whether we have such a gate and returns a unique name for it
+/// since custom gates are implemented in python, this is a heavy python-space function
+pub fn recognize_custom_operation(op: &PackedOperation, name: &String) -> PyResult<Option<String>> {
+    Python::attach(|py| {
```

Comment:
For this and other functions in the file - how about explicitly expect a Python token in the function signature and have the caller attach? This will reduce the nesting level and will make the function signature better capture the intent of the function being Python-dependent.